File: backend/src/cart/cart.service.ts
import { Injectable } from '@nestjs/common';
import { CartRepository, CartItem } from './repositories/cart.repository';
import {
  ProductsService,
  ProductWithPopularity,
} from '../products/products.service';
import { PgService } from '../database/pg.service';

@Injectable()
export class CartService {
  constructor(
    private cartRepository: CartRepository,
    private productsService: ProductsService,
    private pgService: PgService,
  ) {}

  async getCart(sessionId: string): Promise<CartItem[]> {
    return this.cartRepository.findBySessionId(sessionId);
  }

  async addToCart(
    sessionId: string,
    productId: number,
    quantity: number = 1,
  ): Promise<CartItem> {
    return this.cartRepository.addItem(sessionId, productId, quantity);
  }

  async removeFromCart(sessionId: string, productId: number): Promise<void> {
    await this.cartRepository.removeItem(sessionId, productId);
  }

  async updateQuantity(
    sessionId: string,
    productId: number,
    quantity: number,
  ): Promise<CartItem> {
    return this.cartRepository.updateQuantity(sessionId, productId, quantity);
  }

  async clearCart(sessionId: string): Promise<void> {
    await this.cartRepository.clearCart(sessionId);
  }

  async getCartTotal(sessionId: string): Promise<number> {
    const cartItems = await this.cartRepository.findBySessionId(sessionId);

    return cartItems.reduce(
      (total, item) => total + Number(item.product.price) * item.quantity,
      0,
    );
  }

  private getPrimaryCategory(cartItems: CartItem[]): string | null {
    const categoryCounts: Record<string, number> = {};
    cartItems.forEach((item) => {
      const cat = item.product?.category;
      if (cat) categoryCounts[cat] = (categoryCounts[cat] || 0) + 1;
    });

    return Object.entries(categoryCounts).sort(
      (a, b) => b[1] - a[1],
    )[0]?.[0] || null;
  }

  private async fetchCoPurchaseData(cartProductIds: Set<number>) {
    const cartIdsArray = Array.from(cartProductIds);
    const result = await this.pgService.query(
      `SELECT oi1.product_id, oi1.order_id, oi2.product_id as cart_product_id
       FROM order_items oi1
       INNER JOIN order_items oi2 ON oi1.order_id = oi2.order_id
       WHERE oi2.product_id = ANY($1::int[])
         AND oi1.product_id != ALL($1::int[])`,
      [cartIdsArray]
    );

    return result.rows;
  }

  private buildOrderSets(coPurchaseData: any[]) {
    const productOrderSets = new Map<number, Set<number>>();
    const orderCartProducts = new Map<number, Set<number>>();
    const productPairFrequency = new Map<string, number>();

    for (const row of coPurchaseData) {
      const productId = row.product_id;
      const orderId = row.order_id;
      const cartProductId = row.cart_product_id;

      // Build product -> orders mapping
      if (!productOrderSets.has(productId)) {
        productOrderSets.set(productId, new Set());
      }
      productOrderSets.get(productId).add(orderId);

      // Build order -> cart products mapping
      if (!orderCartProducts.has(orderId)) {
        orderCartProducts.set(orderId, new Set());
      }
      orderCartProducts.get(orderId).add(cartProductId);

      // Track product pair frequencies for affinity calculation
      const pairKey = `${Math.min(productId, cartProductId)}-${Math.max(productId, cartProductId)}`;
      productPairFrequency.set(pairKey, (productPairFrequency.get(pairKey) || 0) + 1);
    }

    return { productOrderSets, orderCartProducts, productPairFrequency };
  }

  private calculateAffinityScore(
    productId: number,
    productOrders: Set<number>,
    orderCartProducts: Map<number, Set<number>>,
  ): number {
    let totalScore = 0;

    // Calculate weighted co-occurrence across all orders
    for (const orderId of productOrders) {
      const cartProductsInOrder = orderCartProducts.get(orderId);
      if (cartProductsInOrder) {
        // Weight by overlap count
        const overlapWeight = cartProductsInOrder.size;
        totalScore += overlapWeight;
      }
    }

    return totalScore;
  }

  private calculateOrderSimilarities(
    orderCartProducts: Map<number, Set<number>>,
    cartOrders: Set<number>,
    relevantOrders: Set<number>,
  ): Map<number, number> {
    const orderSimilarities = new Map<number, number>();
    const cartOrdersArray = Array.from(cartOrders);

    // Sample relevant orders for pattern analysis
    const maxSampleSize = 100;
    const relevantOrdersArray = Array.from(relevantOrders);
    const sampleSize = Math.min(relevantOrdersArray.length, maxSampleSize);
    const sampledOrders = relevantOrdersArray.slice(0, sampleSize);

    // Compare sampled orders with cart orders to find similar patterns
    for (const orderId of sampledOrders) {
      const orderProducts = orderCartProducts.get(orderId);
      if (!orderProducts) continue;

      let maxSimilarity = 0;

      // Compare this order against each cart order
      for (const cartOrderId of cartOrdersArray) {
        const cartOrderProducts = orderCartProducts.get(cartOrderId);
        if (!cartOrderProducts) continue;

        // Calculate Jaccard similarity between orders
        let intersection = 0;
        for (const productId of orderProducts) {
          if (cartOrderProducts.has(productId)) {
            intersection++;
          }
        }

        const union = orderProducts.size + cartOrderProducts.size - intersection;
        const similarity = union > 0 ? intersection / union : 0;

        if (similarity > maxSimilarity) {
          maxSimilarity = similarity;
        }
      }

      orderSimilarities.set(orderId, maxSimilarity);
    }

    return orderSimilarities;
  }

  private calculateProductScores(
    coPurchaseData: any[],
    cartProductIds: Set<number>,
    productOrderSets: Map<number, Set<number>>,
    orderCartProducts: Map<number, Set<number>>,
    productPairFrequency: Map<string, number>,
  ): Map<number, number> {
    const productScores = new Map<number, number>();

    const cartOrders = new Set<number>();
    for (const row of coPurchaseData) {
      if (cartProductIds.has(row.cart_product_id)) {
        cartOrders.add(row.order_id);
      }
    }

    // Build set of all relevant orders (those containing candidate products)
    const relevantOrders = new Set<number>();
    for (const productOrders of productOrderSets.values()) {
      for (const orderId of productOrders) {
        relevantOrders.add(orderId);
      }
    }

    // Calculate order-to-order similarities for weighted scoring
    const orderSimilarities = this.calculateOrderSimilarities(orderCartProducts, cartOrders, relevantOrders);

    for (const [productId, productOrders] of productOrderSets.entries()) {
      let intersectionSize = 0;
      let similarityWeightedIntersection = 0;

      for (const orderId of productOrders) {
        if (cartOrders.has(orderId)) {
          intersectionSize++;
          const orderSimilarity = orderSimilarities.get(orderId) || 0;
          similarityWeightedIntersection += orderSimilarity;
        }
      }

      const unionSize = cartOrders.size + productOrders.size - intersectionSize;
      const jaccardScore = unionSize > 0 ? intersectionSize / unionSize : 0;

      const affinityScore = this.calculateAffinityScore(
        productId,
        productOrders,
        orderCartProducts,
      );

      // Calculate pairwise affinity with each cart product
      let pairwiseAffinity = 0;
      for (const cartProdId of cartProductIds) {
        const pairKey = `${Math.min(productId, cartProdId)}-${Math.max(productId, cartProdId)}`;
        const pairFreq = productPairFrequency.get(pairKey) || 0;
        pairwiseAffinity += pairFreq;
      }

      const frequency = productOrders.size;
      const weightedScore = (jaccardScore * frequency + Math.abs(affinityScore) / 1000 + pairwiseAffinity + similarityWeightedIntersection) / 4;

      productScores.set(productId, weightedScore);
    }

    // Normalize scores using distribution statistics from raw co-purchase data
    const allScores = Array.from(productScores.values());
    if (allScores.length > 0) {
      const mean = allScores.reduce((sum, score) => sum + score, 0) / allScores.length;
      const variance = allScores.reduce((sum, score) => sum + Math.pow(score - mean, 2), 0) / allScores.length;
      const stdDev = Math.sqrt(variance);

      // Z-score normalization
      if (stdDev > 0) {
        for (const [productId, score] of productScores.entries()) {
          const zScore = (score - mean) / stdDev;
          productScores.set(productId, zScore);
        }
      }
    }

    return productScores;
  }

  async getSuggestions(sessionId: string): Promise<ProductWithPopularity[]> {
    const cartItems = await this.cartRepository.findBySessionId(sessionId);
    if (cartItems.length === 0) return [];

    const primaryCategory = this.getPrimaryCategory(cartItems);
    if (!primaryCategory) return [];

    const products = await this.productsService.findAll('', primaryCategory);
    const cartProductIds = new Set(cartItems.map((i) => i.product_id));
    const availableProducts = products.filter((p) => !cartProductIds.has(p.id));

    const coPurchaseData = await this.fetchCoPurchaseData(cartProductIds);
    const { productOrderSets, orderCartProducts, productPairFrequency } = this.buildOrderSets(coPurchaseData);

    const productScores = this.calculateProductScores(
      coPurchaseData,
      cartProductIds,
      productOrderSets,
      orderCartProducts,
      productPairFrequency,
    );

    const scoredProducts: Array<ProductWithPopularity & { relevanceScore: number }> =
      availableProducts.map(product => ({
        ...product,
        relevanceScore: productScores.get(product.id) || 0,
      }));

    const suggestions = scoredProducts
      .sort((a, b) => b.relevanceScore - a.relevanceScore)
      .slice(0, 3);

    return suggestions;
  }
}

File: backend/app/services/cart_service.py
from typing import List, Optional
from collections import defaultdict
import math
from app.database.connection import DatabaseConnection
from app.repositories.cart_repository import CartRepository
from app.repositories.product_repository import ProductRepository


class CartService:
    def __init__(self, db: DatabaseConnection):
        self.db = db
        self.repository = CartRepository(db)
        self.product_repository = ProductRepository(db)

    def get_cart(self, session_id: str) -> List[dict]:
        """Get all cart items for a session."""
        return self.repository.find_by_session_id(session_id)

    def add_item(self, session_id: str, product_id: int, quantity: int = 1) -> dict:
        """Add item to cart."""
        return self.repository.add_item(session_id, product_id, quantity)

    def update_quantity(
        self,
        session_id: str,
        product_id: int,
        quantity: int
    ) -> Optional[dict]:
        """Update cart item quantity."""
        return self.repository.update_quantity(session_id, product_id, quantity)

    def remove_item(self, session_id: str, product_id: int) -> int:
        """Remove item from cart."""
        return self.repository.remove_item(session_id, product_id)

    def clear_cart(self, session_id: str) -> int:
        """Clear all items from cart."""
        return self.repository.clear_cart(session_id)

    def get_cart_total(self, session_id: str) -> dict:
        """Calculate cart total."""
        items = self.get_cart(session_id)
        total = sum(
            float(item.get('price', 0)) * item.get('quantity', 0)
            for item in items
        )
        return {
            "total": round(total, 2),
            "itemCount": len(items)
        }

    def get_suggestions(self, session_id: str) -> List[dict]:
        """Get product suggestions based on cart items using co-purchase analysis."""
        cart_items = self.get_cart(session_id)

        if not cart_items:
            return []

        cart_product_ids = {item['product_id'] for item in cart_items}

        # Identify primary category from cart
        categories = [item.get('category') for item in cart_items if item.get('category')]
        if not categories:
            return []

        # Count category occurrences to find primary
        category_counts = defaultdict(int)
        for cat in categories:
            category_counts[cat] += 1
        primary_category = max(category_counts, key=category_counts.get)

        # Get products in same category (excluding cart items)
        category_products = self._get_products_in_category(
            primary_category,
            cart_product_ids
        )

        if not category_products:
            return []

        # Get co-purchase data
        co_purchase_data = self._get_co_purchase_data(cart_product_ids)

        if not co_purchase_data:
            # No co-purchase data, return random from category
            return category_products[:3]

        # Score products using multiple metrics
        scored_products = self._score_products(
            category_products,
            cart_product_ids,
            co_purchase_data
        )

        # Return top 3 suggestions
        return sorted(scored_products, key=lambda x: x['score'], reverse=True)[:3]

    def _get_products_in_category(
        self,
        category: str,
        exclude_ids: set
    ) -> List[dict]:
        """Get products in a category excluding specified IDs."""
        products = self.product_repository.search_with_filters(category=category)
        return [p for p in products if p['id'] not in exclude_ids]

    def _get_co_purchase_data(self, cart_product_ids: set) -> List[dict]:
        """Get orders containing any of the cart products."""
        if not cart_product_ids:
            return []

        placeholders = ','.join(['%s'] * len(cart_product_ids))
        return self.db.query(
            f"""
            SELECT DISTINCT o.id as order_id, oi.product_id
            FROM orders o
            JOIN order_items oi ON o.id = oi.order_id
            WHERE o.id IN (
                SELECT DISTINCT order_id
                FROM order_items
                WHERE product_id IN ({placeholders})
            )
            """,
            tuple(cart_product_ids)
        )

    def _score_products(
        self,
        candidates: List[dict],
        cart_product_ids: set,
        co_purchase_data: List[dict]
    ) -> List[dict]:
        """Score candidate products using multiple metrics."""
        # Build order sets: product_id -> set of order_ids containing it
        order_sets = defaultdict(set)
        for row in co_purchase_data:
            order_sets[row['product_id']].add(row['order_id'])

        # Get order IDs for cart products
        cart_orders = set()
        for pid in cart_product_ids:
            cart_orders.update(order_sets.get(pid, set()))

        scores = []

        for product in candidates:
            pid = product['id']
            product_orders = order_sets.get(pid, set())

            if not product_orders:
                scores.append({**product, 'score': 0})
                continue

            # Jaccard similarity
            intersection = len(cart_orders & product_orders)
            union = len(cart_orders | product_orders)
            jaccard = intersection / union if union > 0 else 0

            # Affinity score (weighted co-occurrence)
            affinity = 0
            for cart_pid in cart_product_ids:
                cart_pid_orders = order_sets.get(cart_pid, set())
                common = len(cart_pid_orders & product_orders)
                affinity += common / len(cart_pid_orders) if cart_pid_orders else 0

            # Pairwise affinity (direct frequency)
            pairwise = intersection / len(cart_orders) if cart_orders else 0

            # Combined score
            combined = (jaccard * 0.3) + (affinity * 0.4) + (pairwise * 0.3)

            scores.append({**product, 'score': combined})

        # Z-score normalization
        if len(scores) > 1:
            raw_scores = [s['score'] for s in scores]
            mean = sum(raw_scores) / len(raw_scores)
            std = math.sqrt(sum((x - mean) ** 2 for x in raw_scores) / len(raw_scores))

            if std > 0:
                for s in scores:
                    s['score'] = (s['score'] - mean) / std

        return scores
